### ISSUE TITLE
Build problems on macOS

### ISSUE THREAD

**Issue by eschnett** [urbit/urbit#791]
The `urbit-0.4` tarball doesn't build for me. I'm using macOS Sierra, and I have the MacPorts compilers installed. (These provide versions of `gcc` that are actually a GCC, not a link to Clang.)

The following patch to `Makefile` solved my build problems:
```
--- urbit-0.4-orig/Makefile     2016-10-25 12:52:21.000000000 -0400
+++ urbit-0.4/Makefile  2016-12-07 11:29:31.000000000 -0500
@@ -65,7 +65,12 @@
 CURLLIB=$(shell curl-config --libs)

 RM=rm -f
-ifneq ($(UNAME),FreeBSD)
+ifeq ($(UNAME),Darwin)
+CC=/usr/bin/cc
+CXX=/usr/bin/c++
+CXXFLAGS=$(CFLAGS)
+CLD=/usr/bin/c++ $(CFLAGS) -L/usr/local/lib $(OPTLOCALLFLAGS) $(OPENSSLLFLAGS)
+else ifneq ($(UNAME),FreeBSD)
 CC=gcc
 CXX=g++
 CXXFLAGS=$(CFLAGS)
@@ -78,7 +83,7 @@
 endif

 ifeq ($(OS),osx)
-  COSFLAGS=-fno-diagnostics-fixit-info
+  # COSFLAGS=-fno-diagnostics-fixit-info
   CLDOSFLAGS=-bind_at_load
   OSLIBS=-framework CoreServices -framework CoreFoundation
 endif
@@ -109,7 +114,7 @@

 LIBUV_VER=libuv-v1.7.5

-LIBUV_CONFIGURE_OPTIONS=
+LIBUV_CONFIGURE_OPTIONS=CC=$(CC)

 # NOTFORCHECKIN - restore -O3
 #      -DGHETTO \
@@ -139,7 +144,7 @@
         -Wno-unused-parameter \
         -Wno-missing-field-initializers \
         -Wno-strict-aliasing \
-        -Werror
+        -Wno-error
 ifneq ($(OS),bsd)
   CWFLAGS+=-Wno-error=unused-result
 endif
```

* use `/usr/bin/cc` as compiler, since we know that's a good compiler on macOS
* don't use `-Werror` in a release, since newer compilers or headers here lead to warnings that abort a build that otherwise works fine
* tell `libuv` not to use GCC since GCC doesn't get along with Sierra's non-standard system headers

**Comment by juped** [comment]
Inviting comment: can anyone think of any reason to use 'gcc' as CC on any platform?

**Comment by cgyarvin** [comment]
Well, I would certainly use gcc on a gcc-native platform.  When in
Stallmania, do as the Stallmans.  Otherwise...

I haven't taken the Sierra plunge and I am on the homebrew side.  However,
we obviously need to support Sierra etc.  Also, plenty of people have other
reasons to be in the MacPorts world, the gcc world, etc -- although here,
we're definitely getting closer to a personal customization.

I agree with not using -Werror in a release build.

Ray, any other comments on this patch?  It doesn't look to me like it'll
break anything...

**Comment by juped** [comment]
Let me rephrase that for the pedantic: 'cc' (name of a binary) always gets you a C compiler. Can anyone think of any reason to use 'gcc' (name of a binary) as CC (makefile variable) on any platform?

**Comment by cgyarvin** [comment]
Erik, what happens with CC=/usr/bin/cc on your MacPorts build?

**Comment by eschnett** [comment]
@cgyarvin This `CC=/usr/bin/cc` setting (which is the one I select in the patch) explicitly chooses Apple's variant of Clang, i.e. the regular system compiler. Unless you're careful, this is the setting you want. I think setting `CC=cc` works as well, but `CXX=c++` doesn't because MacPorts redirects `c++` to its `g++`. (It doesn't redirect `cc` -- go figure.) So `CXX=/usr/bin/c++` seems required.

I think libuv prefers `gcc` over `cc`, which is wrong for Sierra since `gcc` doesn't understand the system header files. (They contain Apple-specific `#defines` or Objective C code that is not protected via `#ifdef`s; yes, I know, but people complaining to Apple doesn't seem to have helped so far.)

**Comment by cgyarvin** [comment]
Sorry, I thought MacPorts had put in gcc as /usr/bin/cc.

libuv compiles quite comfortably with Clang in my universe.  It sounds like
MacPorts has a strong attachment to gcc (probably because it wants you to
not have to install Xcode?), and this has to be fought more vigorously on
Sierra.

Ray, can you take the patch?

**Comment by juped** [comment]
To clarify further:

I think it would be a good idea to use `/usr/bin/cc` and `/usr/bin/c++` on every platform - the first at least is guaranteed to exist by common standards, I forget if the second is. This is a broader/generalized version of what Erik proposes. I am looking for potential examples of odd platforms where this would cause issues from anyone with portability experience, but I can't think of any. I haven't changed this before just due to general reluctance to change things.

As for `-Werror` this can just go. It can come back in some form if we get cute with the build system and detect whether we're a release or not while building. Downstream packagers hate it.

Thanks for the patch.

**Comment by juped** [comment]
@eschnett I adapted your patch - if you have time, could you check if the version on the `maint-0.4` branch builds on your setup? Thanks.

**Comment by eschnett** [comment]
@juped Compiling works, but linking fails:
```
$ make
    CCLD  bin/urbit
c++: error: unrecognized command line option '-fno-diagnostics-fixit-info'; did you mean '-fno-diagnostics-color'?
make: *** [bin/urbit] Error 1
```

This is after a `git pull`, a `git status` to ensure there are no local changes, and a `make distclean`.

The reason is likely that I'm using MacPorts, and I have
```
$ which c++
/opt/local/bin/c++
$ c++ --version
c++ (MacPorts gcc6 6.2.0_2) 6.2.0
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
which is not Apple's C++ compiler. For this, I'd have to use `/usr/bin/c++`.

**Comment by juped** [comment]
Argh, I missed that line. It should be taken out as well; not going to do a whole release for it since I’ve screwed up enough releases in the past 48 hours :) but I will at least remove it from the repository...

**Comment by juped** [comment]
And it’s on maint-0.4. It’ll be in any future releases, of course. Thanks for the help.

**Comment by juped** [comment]
Closing because it's integrated, thanks!
